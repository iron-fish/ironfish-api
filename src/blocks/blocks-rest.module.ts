/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { UsersModule } from '../users/users.module';
import { BlocksController } from './blocks.controller';
import { BlocksModule } from './blocks.module';

@Module({
  controllers: [BlocksController],
  imports: [BlocksModule, UsersModule],
})
export class BlocksRestModule {}
