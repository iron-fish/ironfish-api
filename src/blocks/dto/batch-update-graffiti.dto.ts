/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { ApiProperty } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import {
  ArrayMaxSize,
  ArrayMinSize,
  IsArray,
  ValidateNested,
} from 'class-validator';
import { UpdateGraffitiDto } from './update-graffiti.dto';

export class BatchUpdateGraffitiDto {
  @IsArray()
  @ApiProperty({ description: 'hash + graffiti array' })
  @ArrayMinSize(1)
  @ArrayMaxSize(200)
  @ValidateNested({ each: true })
  @Type(() => UpdateGraffitiDto)
  readonly updates!: UpdateGraffitiDto[];
}
