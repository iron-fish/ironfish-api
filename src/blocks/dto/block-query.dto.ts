/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { ApiPropertyOptional } from '@nestjs/swagger';
import { Transform, TransformFnParams, Type } from 'class-transformer';
import {
  IsDefined,
  IsInt,
  IsOptional,
  IsString,
  Min,
  ValidateIf,
} from 'class-validator';
import { stringToBoolean } from '../../common/utils/boolean';

export class BlockQueryDto {
  @ApiPropertyOptional({ description: 'Block hash' })
  @ValidateIf((o: BlockQueryDto) => o.sequence === undefined)
  @IsDefined({
    message: '"hash" or "sequence" required to query for single block',
  })
  @IsString()
  readonly hash?: string;

  @ApiPropertyOptional({ description: 'Block sequence' })
  @ValidateIf((o: BlockQueryDto) => o.hash === undefined)
  @IsDefined({
    message: '"hash" or "sequence" required to query for single block',
  })
  @Min(1)
  @IsInt()
  @Type(() => Number)
  readonly sequence?: number;

  @ApiPropertyOptional({
    description: 'Whether or not to include transactions in the response',
  })
  @IsOptional()
  @Transform(({ value }: TransformFnParams) => stringToBoolean(value))
  readonly with_transactions?: boolean;
}
