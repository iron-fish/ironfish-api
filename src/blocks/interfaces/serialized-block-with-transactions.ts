/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { SerializedTransaction } from '../../transactions/interfaces/serialized-transaction';
import { SerializedBlock } from './serialized-block';

export interface SerializedBlockWithTransactions extends SerializedBlock {
  transactions: SerializedTransaction[];
}
