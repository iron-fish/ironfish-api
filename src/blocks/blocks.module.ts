/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { ApiConfigModule } from '../api-config/api-config.module';
import { EventsModule } from '../events/events.module';
import { PrismaModule } from '../prisma/prisma.module';
import { UsersModule } from '../users/users.module';
import { BlocksService } from './blocks.service';

@Module({
  exports: [BlocksService],
  imports: [ApiConfigModule, EventsModule, PrismaModule, UsersModule],
  providers: [BlocksService],
})
export class BlocksModule {}
