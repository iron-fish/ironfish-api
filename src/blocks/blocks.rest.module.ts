/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { BlocksTransactionsLoaderModule } from '../blocks-transactions-loader/blocks-transactions-loader.module';
import { UsersModule } from '../users/users.module';
import { BlocksController } from './blocks.controller';
import { BlocksModule } from './blocks.module';

@Module({
  controllers: [BlocksController],
  imports: [BlocksModule, BlocksTransactionsLoaderModule, UsersModule],
})
export class BlocksRestModule {}
