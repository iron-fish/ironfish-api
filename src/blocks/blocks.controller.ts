/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Body,
  Controller,
  Get,
  HttpStatus,
  Post,
  Query,
  Res,
  UnprocessableEntityException,
  UseGuards,
  ValidationPipe,
} from '@nestjs/common';
import { Response } from 'express';
import { ApiKeyGuard } from '../auth/guards/api-key.guard';
import { List } from '../common/interfaces/list';
import { BlocksService } from './blocks.service';
import { BlocksQueryDto } from './dto/blocks-query.dto';
import { DisconnectBlocksDto } from './dto/disconnect-blocks.dto';
import { UpsertBlocksDto } from './dto/upsert-blocks.dto';
import { Block } from '.prisma/client';

@Controller('blocks')
export class BlocksController {
  constructor(private readonly blocksService: BlocksService) {}

  @Post()
  @UseGuards(ApiKeyGuard)
  async bulkUpsert(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    blocks: UpsertBlocksDto,
  ): Promise<List<Block>> {
    return {
      data: await this.blocksService.bulkUpsert(blocks),
    };
  }

  @Get('head')
  async head(): Promise<Block> {
    return this.blocksService.head();
  }

  @Get()
  async list(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { sequence_gte: sequenceGte, sequence_lt: sequenceLt }: BlocksQueryDto,
  ): Promise<List<Block>> {
    const maxBlocksToReturn = 1000;
    if (sequenceGte >= sequenceLt) {
      throw new UnprocessableEntityException(
        `'sequence_gte' must be strictly less than 'sequence_lt'.`,
      );
    }
    if (sequenceLt - sequenceGte > maxBlocksToReturn) {
      throw new UnprocessableEntityException(
        `Range is too long. Max sequence difference is ${maxBlocksToReturn}.`,
      );
    }
    return {
      data: await this.blocksService.list(sequenceGte, sequenceLt),
    };
  }

  @Post('disconnect')
  @UseGuards(ApiKeyGuard)
  async disconnect(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { sequence_gt: sequenceGt }: DisconnectBlocksDto,
    @Res() res: Response,
  ): Promise<void> {
    await this.blocksService.disconnectAfter(sequenceGt);
    res.sendStatus(HttpStatus.OK);
  }
}
