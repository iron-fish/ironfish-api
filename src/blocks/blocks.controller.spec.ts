/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import { Block } from '@prisma/client';
import faker from 'faker';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import { BlocksDailyService } from '../blocks-daily/blocks-daily.service';
import { MetricsGranularity } from '../common/enums/metrics-granularity';
import { PrismaService } from '../prisma/prisma.service';
import { bootstrapTestApp } from '../test/test-app';
import { BlocksService } from './blocks.service';
import { UpsertBlocksDto } from './dto/upsert-blocks.dto';
import { BlockOperation } from './enums/block-operation';
import { SerializedBlockWithTransactions } from './interfaces/serialized-block-with-transactions';

const API_KEY = 'test';

describe('BlocksController', () => {
  let app: INestApplication;
  let blocksDailyService: BlocksDailyService;
  let blocksService: BlocksService;
  let prisma: PrismaService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    blocksDailyService = app.get(BlocksDailyService);
    blocksService = app.get(BlocksService);
    prisma = app.get(PrismaService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  const seedBlock = async () => {
    const hash = uuid();
    const sequence = faker.datatype.number();

    const block = await prisma.block.create({
      data: {
        hash,
        main: true,
        sequence,
        timestamp: new Date(),
        transactions_count: 0,
        graffiti: uuid(),
        previous_block_hash: uuid(),
        network_version: 0,
        size: faker.datatype.number(),
        difficulty_temporary: faker.datatype.number(),
      },
    });

    return { block };
  };

  describe('POST /blocks', () => {
    describe('with a missing api key', () => {
      it('returns a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/blocks`)
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with missing arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/blocks`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ blocks: [{}] })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with too many blocks', () => {
      it('returns a 422', async () => {
        const blocks = [];
        for (let i = 0; i < 3001; i++) {
          blocks.push({
            hash: uuid(),
            difficulty: faker.datatype.number(),
            type: BlockOperation.CONNECTED,
            sequence: faker.datatype.number(),
            timestamp: new Date(),
            transactions_count: 0,
            graffiti: uuid(),
            previous_block_hash: uuid(),
            size: faker.datatype.number(),
            difficulty_temporary: faker.datatype.number(),
            transactions: [
              {
                hash: uuid(),
                fee: faker.datatype.number(),
                size: faker.datatype.number(),
                notes: [{ commitment: uuid() }],
                spends: [{ nullifier: uuid() }],
              },
            ],
          });
        }
        const payload: UpsertBlocksDto = { blocks };

        const { body } = await request(app.getHttpServer())
          .post(`/blocks`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send(payload)
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a valid payload', () => {
      it('upserts blocks', async () => {
        const payload: UpsertBlocksDto = {
          blocks: [
            {
              hash: uuid(),
              difficulty: faker.datatype.number(),
              type: BlockOperation.CONNECTED,
              sequence: faker.datatype.number(),
              timestamp: new Date(),
              graffiti: uuid(),
              previous_block_hash: uuid(),
              size: faker.datatype.number(),
              transactions: [
                {
                  hash: uuid(),
                  fee: faker.datatype.number(),
                  size: faker.datatype.number(),
                  notes: [{ commitment: uuid() }],
                  spends: [{ nullifier: uuid() }],
                },
              ],
            },
          ],
        };
        const block = payload.blocks[0];
        const { body } = await request(app.getHttpServer())
          .post(`/blocks`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send(payload)
          .expect(HttpStatus.CREATED);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          hash: block.hash,
          difficulty: block.difficulty,
          sequence: block.sequence,
          timestamp: block.timestamp.toISOString(),
          transactions_count: 1,
          previous_block_hash: block.previous_block_hash,
          size: block.size,
        });
      });
    });
  });

  describe('GET /blocks/head', () => {
    it('returns the heaviest block', async () => {
      const { body } = await request(app.getHttpServer())
        .get('/blocks/head')
        .expect(HttpStatus.OK);

      expect(body).toMatchObject({
        id: expect.any(Number),
        main: true,
      });
    });
  });

  describe('GET /blocks/status', () => {
    it('returns information about the main chain', async () => {
      const { body } = await request(app.getHttpServer())
        .get('/blocks/status')
        .expect(HttpStatus.OK);

      expect(body).toMatchObject({
        chain_height: expect.any(Number),
        percentage_marked: expect.any(Number),
        unique_graffiti: expect.any(Number),
      });
    });
  });

  describe('GET /blocks', () => {
    describe('with no query parameters', () => {
      it('returns a list of blocks in descending order', async () => {
        for (let i = 0; i < 10; i++) {
          const hash = uuid();
          const testGraffiti = `graffiti ${i}`;
          await prisma.block.create({
            data: {
              hash,
              main: true,
              sequence: i,
              timestamp: new Date(),
              transactions_count: 0,
              graffiti: testGraffiti,
              previous_block_hash: uuid(),
              network_version: 0,
              size: faker.datatype.number(),
              difficulty_temporary: faker.datatype.number(),
            },
          });
        }
        const { body } = await request(app.getHttpServer())
          .get('/blocks')
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThanOrEqual(10);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          hash: expect.any(String),
          difficulty: expect.any(Number),
          sequence: expect.any(Number),
          timestamp: expect.any(String),
          transactions_count: expect.any(Number),
          previous_block_hash: expect.any(String),
          size: expect.any(Number),
        });
        expect(((data as unknown[])[0] as Block).id).toBeGreaterThan(
          ((data as unknown[])[1] as Block).id,
        );
      });

      describe('with main chain parameter set to false', () => {
        it('returns only non-main chain blocks', async () => {
          for (let i = 0; i < 10; i++) {
            const hash = uuid();
            const testGraffiti = `graffiti ${i}`;
            await prisma.block.create({
              data: {
                hash,
                main: i < 5,
                sequence: i,
                timestamp: new Date(),
                transactions_count: 0,
                graffiti: testGraffiti,
                previous_block_hash: uuid(),
                network_version: 0,
                size: faker.datatype.number(),
                difficulty_temporary: faker.datatype.number(),
              },
            });
          }

          const { body } = await request(app.getHttpServer())
            .get('/blocks')
            .query({ main: false })
            .expect(HttpStatus.OK);

          const { data } = body;
          expect((data as unknown[]).length).toBeGreaterThanOrEqual(5);
          expect((data as unknown[])[0]).toMatchObject({
            id: expect.any(Number),
            hash: expect.any(String),
            difficulty: expect.any(Number),
            main: false,
            sequence: expect.any(Number),
            timestamp: expect.any(String),
            transactions_count: expect.any(Number),
            previous_block_hash: expect.any(String),
            size: expect.any(Number),
          });
        });
      });
    });

    describe('with invalid sequence_gte and sequence_lt parameters', () => {
      describe('when sequence_gte and sequence_lt are not at least 1', () => {
        it('returns a 422', async () => {
          const { body } = await request(app.getHttpServer())
            .get('/blocks')
            .query({ sequence_gte: -1, sequence_lt: -1 })
            .expect(HttpStatus.UNPROCESSABLE_ENTITY);

          expect(body).toMatchSnapshot();
        });
      });

      describe('when sequence_gte > sequence_lt', () => {
        it('returns a 422', async () => {
          const { body } = await request(app.getHttpServer())
            .get('/blocks')
            .query({ sequence_gte: 2, sequence_lt: 1 })
            .expect(HttpStatus.UNPROCESSABLE_ENTITY);

          expect(body).toMatchSnapshot();
        });
      });

      describe('when the range is too long', () => {
        it('returns a 422', async () => {
          const { body } = await request(app.getHttpServer())
            .get('/blocks')
            .query({ sequence_gte: 1, sequence_lt: 1002 })
            .expect(HttpStatus.UNPROCESSABLE_ENTITY);

          expect(body).toMatchSnapshot();
        });
      });
    });

    describe('with a valid sequence range', () => {
      it('returns blocks within the range', async () => {
        // Seed some blocks
        for (let i = 0; i < 10; i++) {
          const hash = uuid();
          const testGraffiti = `graffiti ${i}`;
          await prisma.block.create({
            data: {
              hash,
              main: true,
              sequence: i,
              timestamp: new Date(),
              transactions_count: 0,
              graffiti: testGraffiti,
              previous_block_hash: uuid(),
              network_version: 0,
              size: faker.datatype.number(),
              difficulty_temporary: faker.datatype.number(),
            },
          });
        }

        const { body } = await request(app.getHttpServer())
          .get('/blocks')
          .query({ sequence_gte: 1, sequence_lt: 420 })
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          hash: expect.any(String),
          difficulty: expect.any(Number),
          sequence: expect.any(Number),
          timestamp: expect.any(String),
          transactions_count: expect.any(Number),
          previous_block_hash: expect.any(String),
          size: expect.any(Number),
        });
      });
    });

    describe('with search parameters', () => {
      describe('when string is a valid partial hash', () => {
        it('returns block(s) with a match', async () => {
          const testBlockHash = uuid();
          const testSequence = faker.datatype.number();
          const testGraffiti = `graffiti ${testSequence}`;
          await prisma.block.create({
            data: {
              hash: testBlockHash,
              main: true,
              sequence: testSequence,
              timestamp: new Date(),
              transactions_count: 0,
              graffiti: testGraffiti,
              previous_block_hash: uuid(),
              network_version: 0,
              size: faker.datatype.number(),
              difficulty_temporary: faker.datatype.number(),
            },
          });

          const { body } = await request(app.getHttpServer())
            .get('/blocks')
            .query({ search: testBlockHash })
            .expect(HttpStatus.OK);

          const { data } = body;
          expect((data as unknown[]).length).toBeGreaterThan(0);
          expect((data as unknown[])[0]).toMatchObject({
            id: expect.any(Number),
            hash: testBlockHash,
            difficulty: expect.any(Number),
            sequence: expect.any(Number),
            timestamp: expect.any(String),
            transactions_count: expect.any(Number),
            previous_block_hash: expect.any(String),
            size: expect.any(Number),
          });
        });
      });

      describe('when string is a valid sequence', () => {
        it('returns block(s) with a match', async () => {
          const testBlockHash = uuid();
          const testSequence = 12345;
          const testGraffiti = `graffiti ${testSequence}`;
          await prisma.block.create({
            data: {
              hash: testBlockHash,
              main: true,
              sequence: testSequence,
              timestamp: new Date(),
              transactions_count: 0,
              graffiti: testGraffiti,
              previous_block_hash: uuid(),
              network_version: 0,
              size: faker.datatype.number(),
              difficulty_temporary: faker.datatype.number(),
            },
          });

          const { body } = await request(app.getHttpServer())
            .get('/blocks')
            .query({ search: testSequence })
            .expect(HttpStatus.OK);

          const { data } = body;
          expect((data as unknown[]).length).toBeGreaterThan(0);
          expect((data as unknown[])[0]).toMatchObject({
            id: expect.any(Number),
            hash: expect.any(String),
            difficulty: expect.any(Number),
            sequence: testSequence,
            timestamp: expect.any(String),
            transactions_count: expect.any(Number),
            previous_block_hash: expect.any(String),
            size: expect.any(Number),
          });
        });
      });
    });

    describe('with a transaction ID', () => {
      it('returns block(s) that contain said transaction', async () => {
        const notes = [{ commitment: uuid() }];
        const spends = [{ nullifier: uuid() }];
        const transaction = await prisma.transaction.create({
          data: {
            hash: uuid(),
            network_version: 0,
            fee: faker.datatype.number(),
            size: faker.datatype.number(),
            notes,
            spends,
          },
        });
        for (let i = 0; i < 9; i++) {
          const { block } = await seedBlock();
          await prisma.blockTransaction.create({
            data: {
              block_id: block.id,
              transaction_id: transaction.id,
            },
          });
        }

        const { body } = await request(app.getHttpServer())
          .get('/blocks')
          .query({ transaction_id: transaction.id, with_transactions: true })
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        for (const serializedBlock of data as SerializedBlockWithTransactions[]) {
          for (const tx of serializedBlock.transactions) {
            expect(tx.id).toBe(transaction.id);
          }
        }
      });
    });
  });

  describe('GET /blocks/find', () => {
    describe('with a valid hash', () => {
      it('returns the block with the correct hash', async () => {
        const testBlockHash = uuid();
        const testSequence = faker.datatype.number();
        const testGraffiti = `graffiti ${testSequence}`;
        await prisma.block.create({
          data: {
            hash: testBlockHash,
            main: true,
            sequence: faker.datatype.number(),
            timestamp: new Date(),
            transactions_count: 0,
            graffiti: testGraffiti,
            previous_block_hash: uuid(),
            network_version: 0,
            size: faker.datatype.number(),
            difficulty_temporary: faker.datatype.number(),
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/blocks/find')
          .query({ hash: testBlockHash })
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          id: expect.any(Number),
          hash: testBlockHash,
          difficulty: expect.any(Number),
          sequence: expect.any(Number),
          timestamp: expect.any(String),
          transactions_count: expect.any(Number),
          previous_block_hash: expect.any(String),
          size: expect.any(Number),
        });
      });
    });

    describe('with a valid sequence', () => {
      it('returns the block with the correct sequence', async () => {
        const hash = uuid();
        const testBlockSequence = faker.datatype.number();
        const testGraffiti = `graffiti ${testBlockSequence}`;
        await prisma.block.create({
          data: {
            hash,
            main: true,
            sequence: testBlockSequence,
            timestamp: new Date(),
            transactions_count: 0,
            graffiti: testGraffiti,
            previous_block_hash: uuid(),
            network_version: 0,
            size: faker.datatype.number(),
            difficulty_temporary: faker.datatype.number(),
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/blocks/find')
          .query({ sequence: testBlockSequence })
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          id: expect.any(Number),
          hash: expect.any(String),
          difficulty: expect.any(Number),
          sequence: testBlockSequence,
          timestamp: expect.any(String),
          transactions_count: expect.any(Number),
          previous_block_hash: expect.any(String),
          size: expect.any(Number),
        });
      });
    });

    describe('with neither a matching hash nor sequence', () => {
      it('returns a 404', async () => {
        const hash = uuid();
        const sequence = faker.datatype.number();
        const testGraffiti = `graffiti ${sequence}`;

        await prisma.block.create({
          data: {
            hash,
            main: true,
            sequence,
            timestamp: new Date(),
            transactions_count: 0,
            graffiti: testGraffiti,
            previous_block_hash: uuid(),
            network_version: 0,
            size: faker.datatype.number(),
            difficulty_temporary: faker.datatype.number(),
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/blocks/find')
          .query({ hash: uuid(), sequence: faker.datatype.number() })
          .expect(HttpStatus.NOT_FOUND);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with neither a valid hash nor sequence', () => {
      it('returns a 422', async () => {
        const hash = uuid();
        const sequence = faker.datatype.number();
        const testGraffiti = `graffiti ${sequence}`;
        await prisma.block.create({
          data: {
            hash,
            main: true,
            sequence,
            timestamp: new Date(),
            transactions_count: 0,
            graffiti: testGraffiti,
            previous_block_hash: uuid(),
            network_version: 0,
            size: faker.datatype.number(),
            difficulty_temporary: faker.datatype.number(),
          },
        });

        const { body } = await request(app.getHttpServer())
          .get('/blocks/find')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });
  });

  describe('POST /blocks/disconnect', () => {
    describe('with a missing api key', () => {
      it('returns a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/blocks/disconnect`)
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with missing arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/blocks/disconnect`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('wtih a valid sequence', () => {
      it('disconnects blocks after the sequence', async () => {
        const disconnectAfter = jest.spyOn(blocksService, 'disconnectAfter');
        const sequenceGt = 2;
        await request(app.getHttpServer())
          .post(`/blocks/disconnect`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ sequence_gt: sequenceGt })
          .expect(HttpStatus.OK);

        expect(disconnectAfter).toHaveBeenCalledWith(sequenceGt);
      });
    });
  });

  describe('GET /blocks/metrics', () => {
    describe('with missing arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/blocks/metrics')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with invalid granularity', () => {
      it('returns a 422', async () => {
        const start = '2021-06-01T00:00:00.000Z';
        const end = '2021-07-01T00:00:00.000Z';
        const { body } = await request(app.getHttpServer())
          .get('/blocks/metrics')
          .query({ start, end, granularity: MetricsGranularity.LIFETIME })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with start after end', () => {
      it('returns a 422', async () => {
        const start = new Date().toISOString();
        const end = new Date(Date.now() - 1).toISOString();
        const { body } = await request(app.getHttpServer())
          .get('/blocks/metrics')
          .query({
            start,
            end,
            granularity: MetricsGranularity.DAY,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a time range longer than the supported range', () => {
      it('returns a 422', async () => {
        const start = '2021-06-01T00:00:00.000Z';
        const end = '2021-10-01T00:00:00.000Z';
        const { body } = await request(app.getHttpServer())
          .get('/blocks/metrics')
          .query({
            start,
            end,
            granularity: MetricsGranularity.DAY,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a valid request', () => {
      it('returns daily snapshots', async () => {
        const start = '2021-06-01T00:00:00.000Z';
        const end = '2021-07-01T00:00:00.000Z';
        await blocksDailyService.upsert(prisma, {
          averageBlockTimeMs: 0,
          averageDifficultyMillis: 0,
          blocksCount: 0,
          blocksWithGraffitiCount: 0,
          chainSequence: 0,
          cumulativeUniqueGraffiti: 0,
          date: new Date(start),
          transactionsCount: 0,
          uniqueGraffiti: 0,
        });

        const { body } = await request(app.getHttpServer())
          .get('/blocks/metrics')
          .query({
            start,
            end,
            granularity: MetricsGranularity.DAY,
          })
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        expect((data as unknown[])[0]).toMatchObject({
          object: 'block_metrics',
          average_block_time_ms: expect.any(Number),
          average_difficulty: expect.any(Number),
          blocks_count: expect.any(Number),
          blocks_with_graffiti_count: expect.any(Number),
          chain_sequence: expect.any(Number),
          cumulative_unique_graffiti: expect.any(Number),
          date: expect.any(String),
          transactions_count: expect.any(Number),
          unique_graffiti_count: expect.any(Number),
        });
      });
    });
  });
});
