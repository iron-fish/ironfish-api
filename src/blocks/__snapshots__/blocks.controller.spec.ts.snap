// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlocksController GET /blocks with invalid sequence_gte and sequence_lt parameters when sequence_gte > sequence_lt returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "'sequence_gte' must be strictly less than 'sequence_lt'.",
  "statusCode": 422,
}
`;

exports[`BlocksController GET /blocks with invalid sequence_gte and sequence_lt parameters when sequence_gte and sequence_lt are not at least 1 returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "sequence_gte must not be less than 1",
    "sequence_lt must not be less than 1",
  ],
  "statusCode": 422,
}
`;

exports[`BlocksController GET /blocks with invalid sequence_gte and sequence_lt parameters when the range is too long returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "Range is too long. Max sequence difference is 1000.",
  "statusCode": 422,
}
`;

exports[`BlocksController GET /blocks/find with neither a matching hash nor sequence returns a 404 1`] = `
Object {
  "message": "Not Found",
  "statusCode": 404,
}
`;

exports[`BlocksController GET /blocks/find with neither a valid hash nor sequence returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "\\"hash\\" or \\"sequence\\" required to query for single block",
    "hash must be a string",
    "\\"hash\\" or \\"sequence\\" required to query for single block",
    "sequence must be an integer number",
    "sequence must not be less than 1",
  ],
  "statusCode": 422,
}
`;

exports[`BlocksController POST /blocks with a missing api key returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`BlocksController POST /blocks with missing arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "blocks.0.hash must be a string",
    "blocks.0.sequence must be an integer number",
    "blocks.0.difficulty must be an integer number",
    "blocks.0.type must be a valid enum value",
    "blocks.0.timestamp must be a Date instance",
    "blocks.0.transactions_count must be an integer number",
    "blocks.0.graffiti must be a string",
  ],
  "statusCode": 422,
}
`;

exports[`BlocksController POST /blocks with too many blocks returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "blocks must contain not more than 3000 elements",
  ],
  "statusCode": 422,
}
`;

exports[`BlocksController POST /blocks/disconnect with a missing api key returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`BlocksController POST /blocks/disconnect with missing arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "sequence_gt must be an integer number",
    "sequence_gt must not be less than 1",
  ],
  "statusCode": 422,
}
`;
